projects/rx-form-mapper/src/lib/services/form-reader.ts

```diff
@@ -0,0 +1,78 @@
+import { InjectFlags, Injector, Type } from '@angular/core';
+import { AbstractControl, FormArray, FormControl, FormGroup } from '@angular/forms';
+import { CustomControlMapper } from '..';
+import { ControlVisitor, CustomControlMetadata, FormArrayMetadata, FormControlMetadata, FormGroupMetadata, FormMetadata } from '../metadata';
+
+export class FormReader implements ControlVisitor<any> {
+
+	public constructor(private readonly control: AbstractControl, private readonly injector: Injector) {}
+
+	public visitCustomControlMetadata(customControlMetadata: CustomControlMetadata): any {
+		return this.resolveCustomMapper(customControlMetadata.mapper).readForm(this.control);
+	}
+
+	public visitFormArrayMetadata(formArrayMetadata: FormArrayMetadata) {
+		if (!this.control) {
+			return void 0;
+		}
+
+		if (!(this.control instanceof FormArray)) {
+			throw new Error('controller is not FormArray instance');
+		}
+
+		return this.control.controls.map(control => {
+			const subFormReader = this.copyPrototype(control);
+			return formArrayMetadata.itemForm.accept(subFormReader);
+		});
+	}
+
+	public visitFormControlMetadata(formControlMetadata: FormControlMetadata) {
+		if (!this.control) {
+			return void 0;
+		}
+
+		if (!(this.control instanceof FormControl)) {
+			throw new Error('controller is not FormControl instance');
+		}
+
+		return this.control.value;
+	}
+
+	public visitFormGroupMetadata(formGroupMetadata: FormGroupMetadata) {
+		if (!this.control) {
+			return void 0;
```

Comment:
Queste funzioni `visit*` hanno codice ripetuto. Metto questa nota qui, ma vale per tutti. Riesci a rimodularle e centralizzare

Reply:
non saprei, ti viene qualche idea in mente?

l'unico punto che potrei risolvere è il check sul tipo estraendo un metodo privato:

```typescript
private assertControlType(value: any, expected: Type<any>) {
                 if (!(value instanceof expected)) {
			throw new Error('Unexpected variable type. Expected ${expected.name}');
		}
}
```

che ne pensi? altre idee?

Reply:
Perché non sfrutti alcuni pattern della programmazione funzionale?
Semplifica il tuo metodo estraendo i vari controlli in singole funzioni (preferibilmente pure).

In questo modo, dove utilizzi il metodo principale, avrai il tuo dato in pipe a n funzioni le quali applicano controlli o (se non sono pure) mutano il dato